src/protosym/core/sym.py

```diff
@@ -0,0 +1,615 @@
+        Sym(TreeAtom(Wild('a')))
+        """
+        return cls(TreeAtom(Wild(name)))
+
+    @classmethod
+    def new_evaluator(
+        cls: Type[T_sym], name: str, typ: Type[T_val]
+    ) -> SymEvaluator[T_sym, T_val]:
+        """Create a :class:`SymEvaluator` for a :class:`Sym` subclass.
+
+        >>> from protosym.core.sym import Sym, PyFunc1
+        >>> a = Sym.new_wild('a')
+        >>> Integer = Sym.new_atom('Integer', int)
+        >>> one = Integer(1)
+        >>> int_to_float = PyFunc1(float)
+        >>> eval_f64 = Sym.new_evaluator('eval_f64', float)
+        >>> eval_f64
+        eval_f64
+        >>> eval_f64[Integer[a]] = int_to_float(a)
+        >>> eval_f64(one)
+        1.0
+
+        See Also
+        ========
+
+        SymEvaluator
+        """
+        return SymEvaluator(name)
+
+
+class PyFunc:
+    """Base class for wrapping callable Python functions.
+
+    This should not be used directly but rather its subclasses that have
+    specific signatures should be used. The purpose of these different classes
+    is really just so that a type checker can understand what type of function
+    is expected by an :class:`Evaluator` and can check the types when a rule is
+    added. Putting together an :class:`Evaluator` at runtime inherently
+    involves dynamic typing so the checker needs a bit of help to understand
+    what is happening.
+
+    See Also
+    ========
+
+    PyOp1
+    PyOp2
+    PyOpN
+    HeadOp
+    AtomFunc
+    PyFunc1
+    """
+
+    __slots__ = ()
+
+
+class WildCall(Generic[T_sym, T_op]):
+
+    __slots__ = ("op", "args")
+
+    def __init__(self, op: T_op, *args: T_sym):
+        self.op = op
+        self.args = args
+
+
+class PyOp(Generic[T_val], PyFunc):
+    """Base class for PyFuncs T -> T, (T, T) -> T etc.
+
+    See Also
+    ========
+
+    PyFunc
+    """
+
+    __slots__ = ()
+
+
+class PyOp1(PyOp[T_val]):
+    """Wrapper for an unary func T_val -> T_val.
+
+    See Also
+    ========
+
+    PyFunc
+    """
+
+    __slots__ = ("func",)
+
+    def __init__(self, func: Callable[[T_val], T_val]):
+        self.func = func
+
+    def __call__(self: T_op, arg1: T_sym) -> WildCall[T_sym, T_op]:
+        return WildCall(self, arg1)
+
+
+class PyOp2(PyOp[T_val]):
+    """Wrapper for an binary func (T_val, T_val) -> T_val.
+
+    See Also
+    ========
+
+    PyFunc
+    """
+
+    __slots__ = ("func",)
+
+    def __init__(self, func: Callable[[T_val, T_val], T_val]):
+        self.func = func
+
+    def __call__(self: T_op, arg1: T_sym, arg2: T_sym) -> WildCall[T_sym, T_op]:
+        return WildCall(self, arg1, arg2)
+
+
+class PyOpN(PyOp[T_val]):
+    """Wrapper for a sequence func [T_val, ...] -> T_val.
+
+    See Also
+    ========
+
+    PyFunc
+    """
+
+    __slots__ = ("func",)
+
+    def __init__(self, func: Callable[[Sequence[T_val]], T_val]):
+        self.func = func
+
+    def __call__(self: T_op, args: T_sym) -> WildCall[T_sym, T_op]:
+        return WildCall(self, args)
```

Comment:
Should `args` not be type hinted as `Sequence[T_sym]`? And the return `WildCall(self, *args)`?

Reply:
No it shouldn't. This is quite confusing with all these classes and I would like to simplify it. I thought I had a better approach but gave up (postponed) because it was taking too long and I thought I was overthinking things.

Here the rule is:
```python
f64_add = PyOpN[float](math.fsum)
eval_f64[Add(star(a))] = f64_add(a)
```
The class we are looking at here is `PyOpN` so `f64_add` is an instance which wraps the function `fsum` that takes a single iterable argument. When `f64_add` itself is called it will be called with a single argument `a` which is a wild symbol representing the single iterable argument to `fsum`. So the argument is `T_sym` (meaning `a`). In the `WildCall` that argument `a` is like a placeholder and when the actual call to `fsum` happens it will be called with `Iterable[float]` in place of `a`.

It all looks very confusing in the internals but hopefully in the external part the `star(a)` captures the fact that we're mapping from `Add(*args)` to `fsum(args)`:
```python
eval_f64[Add(star(a))] = f64_add(a)
```
I did contemplate implementing `__iter__` so that you could literally write it as
```python
eval_f64[Add(*a)] = f64_add(a)
```
That would be nice but I didn't want to add `__iter__` to all of Sym. If there was a way to keep it local to one class then that could work but currently a wild has the same type as the class that creates it (e.g. `Expr.new_wild('a')` is `Expr`).

Perhaps `f64_sum` is a better name to convey the fact that it takes an iterable like `sum` and `fsum` rather than e.g. `operator.add` or `__add__` which are binary.

Reply:
Thanks for the explanation, that makes it clear now.